Null or UNSET_VALUE subscript should generate an invalid request error

If a user tries the to use "NULL" as a map index, e.g., `e` is a map column and one does
```
SELECT * FROM tag WHERE e[null] = 2 ALLOW FILTERING
```
Cassandra yields a InvalidRequest with the message "Unsupported null map value for column e", but Scylla currently yields an internal Server Error, with the message "std::get: wrong index for variant".

Reproduced by cql-pytest `test_null.py`::`test_map_subscript_null`, and also by translated Cassandra unit test `cassandra_tests/validation/operations/select_test.py`::`testFilteringWithoutIndicesWithCollections`

CC @cvybhu

By the way, another non-crashing error we see in test_map_subscript_null when run with a non-empty table (i.e., when the full test file is run) instead of the usual ""std::get: wrong index for variant":
```
E   cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.17.230.140:9042 datacenter1>: <Error from server: code=0000 [Server error] message="marshaling error: read_simple_bytes - not enough bytes (requested 1610612784, got 40) Backtrace: 0x273729e 0x2737790 0x2737a98 0x159a53f 0x159a294 0x157b9b3 0x157b944 0x158ce2f 0x158febb 0x1561b5f 0x157a658 0x13533de 0x1353129 0x1352593 0x135224e 0x134f9bf 0x1345da1 0x1720cc5 0x1721238 0x1472b48 0x1472a30 0x1e01e72 0x1e016b0 0x1e01224 0x1e00572 0x1e00373 0x1e00219 0x1e000e8 0x1dfda77 0x1dfd8e2 0x25ff634 0x26008a7 0x2618ed5 0x25d88fd /lib64/libpthread.so.0+0x92a4 /lib64/libc.so.6+0x100322
```

--- a/cql3/expr/expression.cc
+++ b/cql3/expr/expression.cc
@@ -177,6 +177,35 @@ static managed_bytes_opt get_value(const column_maybe_subscripted& col, const co
             const auto& data_map = value_cast<map_type_impl::native_type>(deserialized);
             const auto key = evaluate(s->sub, options);
             auto&& key_type = col_type->name_comparator();
+            if (key.is_null()) {
+                // For m[null] return null.
+                // This is different from Cassandra - which treats m[null]
+                // as an invalid request error. But m[null] -> null is more
+                // consistent with our usual null treatement (e.g., both
+                // null[2] and null < 2 return null). It will also allow us
+                // to support non-constant subscripts (e.g., m[a]) where "a"
+                // may be null in some rows and non-null in others, and it's
+                // not an error.
+                return std::nullopt;
+            }
+            if (key.is_unset_value()) {
+                // An m[?] with ? bound to UNSET_VALUE is a invalid query.
+                // We could have detected it earlier while binding, but since
+                // we currently don't, we must protect the following code
+                // which can't work with an UNSET_VALUE. Note that the
+                // placement of this check here means that in an empty table,
+                // where we never need to evaluate the filter expression, this
+                // error will not be detected.
+                throw exceptions::invalid_request_exception(
+                    format("Unsupported unset map key for column {}",
+                        cdef->name_as_text()));
+            }
+            if (key.type != key_type) {
+                // This can't happen, we always verify the index type earlier.
+                throw std::logic_error(
+                    format("Tried to evaluate expression with wrong type for subscript of {}",
+                        cdef->name_as_text()));
+            }
             const auto found = key.view().with_linearized([&] (bytes_view key_bv) {
                 using entry = std::pair<data_value, data_value>;
                 return std::find_if(data_map.cbegin(), data_map.cend(), [&] (const entry& element) {

--- a/cql3/expr/expression.cc
+++ b/cql3/expr/expression.cc
@@ -168,7 +168,12 @@ static managed_bytes_opt get_value(const column_maybe_subscripted& col, const co
                         format("Column definition {} does not match any column in the query selection",
                         cdef->name_as_text()));
             }
-            const auto deserialized = cdef->type->deserialize(managed_bytes_view(*data.other_columns[index]));
+            const managed_bytes_opt& serialized = data.other_columns[index];
+            if (!serialized) {
+                // For null[i] we return null.
+                return std::nullopt;
+            }
+            const auto deserialized = cdef->type->deserialize(managed_bytes_view(*serialized));
             const auto& data_map = value_cast<map_type_impl::native_type>(deserialized);
             const auto key = evaluate(s->sub, options);
             auto&& key_type = col_type->name_comparator();

--- a/test/cql-pytest/test_filtering.py
+++ b/test/cql-pytest/test_filtering.py
@@ -14,6 +14,7 @@ import pytest
 from util import new_test_table
 from cassandra.protocol import InvalidRequest
 from cassandra.connection import DRIVER_NAME, DRIVER_VERSION
+from cassandra.query import UNSET_VALUE
 
 # When filtering for "x > 0" or "x < 0", rows with an unset value for x
 # should not match the filter.
@@ -167,3 +168,71 @@ def test_filtering_with_in_relation(cql, test_keyspace, cassandra_bug):
         assert set(res) == set([(1,2,3,4), (3,4,5,6)])
         res = cql.execute(f"select * from {table} where v in (5,7) ALLOW FILTERING")
         assert set(res) == set([(2,3,4,5), (4,5,6,7)])
+
+# Test that subscripts in expressions work as expected. They should only work
+# on map columns, and must have the correct type.
+# This test is a superset of test test_null.py::test_map_subscript_null which
+# tests only the special case of a null subscript.
+# Reproduces #10361
+@pytest.mark.xfail(reason="Issue #10361")
+def test_filtering_with_subscript(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace,
+            "p int, m1 map<int, int>, m2 map<text, text>, s set<int>, PRIMARY KEY (p)") as table:
+        # Check for *errors* in subscript expressions - such as wrong type or
+        # null - with an empty table. This will force the implementation to
+        # check for these errors before actually evaluating the filter
+        # expression - because there will be no rows to filter.
+
+        # A subscript is not allowed on a non-map column (in this case, a set)
+        with pytest.raises(InvalidRequest, match='cannot be used as a map'):
+            cql.execute(f"SELECT p FROM {table} WHERE s[2] = 3 ALLOW FILTERING")
+        # A wrong type is passed for the subscript is not allowed
+        with pytest.raises(InvalidRequest, match='key\(m1\)'):
+            cql.execute(f"select p from {table} where m1['black'] = 2 ALLOW FILTERING")
+        with pytest.raises(InvalidRequest, match='key\(m2\)'):
+            cql.execute(f"select p from {table} where m2[1] = 2 ALLOW FILTERING")
+        # A "null" is not allowed as a key (reproduces #10361)
+        with pytest.raises(InvalidRequest, match='Unsupported null map key for column m1'):
+            cql.execute(f"select p from {table} where m1[null] = 2 ALLOW FILTERING")
+        with pytest.raises(InvalidRequest, match='Unsupported null map key for column m2'):
+            cql.execute(f"select p from {table} where m2[null] = 'hi' ALLOW FILTERING")
+        # Similar to above checks, but using a prepared statement. We can't
+        # cause the driver to send the wrong type to a bound variable, so we
+        # can't check that case unfortunately, but we have a new UNSET_VALUE
+        # case.
+        stmt = cql.prepare(f"select p from {table} where m1[?] = 2 ALLOW FILTERING")
+        with pytest.raises(InvalidRequest, match='Unsupported null map key for column m1'):
+            cql.execute(stmt, [None])
+        with pytest.raises(InvalidRequest, match='Unsupported unset map key for column m1'):
+            cql.execute(stmt, [UNSET_VALUE])
+
+        # Finally, check for sucessful filtering with subscripts. For that we
+        # need to add some data:
+        cql.execute("INSERT INTO "+table+" (p, m1, m2) VALUES (1, {1:2, 3:4}, {'dog':'cat', 'hi':'hello'})")
+        cql.execute("INSERT INTO "+table+" (p, m1, m2) VALUES (2, {2:3, 4:5}, {'man':'woman', 'black':'white'})")
+        res = cql.execute(f"select p from {table} where m1[1] = 2 ALLOW FILTERING")
+        assert list(res) == [(1,)]
+        res = cql.execute(f"select p from {table} where m2['black'] = 'white' ALLOW FILTERING")
+        assert list(res) == [(2,)]
+        res = cql.execute(stmt, [1])
+        assert list(res) == [(1,)]
+
+        # Try again the null-key request (reproduces #10361) that we did
+        # earlier when there was no data in the table. Now there is, and
+        # the scan brings up several rows, it may exercise different code
+        # paths.
+        with pytest.raises(InvalidRequest, match='Unsupported null map key for column m1'):
+            cql.execute(f"select p from {table} where m1[null] = 2 ALLOW FILTERING")
+
+# Beyond the tests of map subscript expressions above, also test what happens
+# when the expression is fine (e.g., m[2] = 3) but the *data* itself is null.
+# We used to have a bug there where we attempted to incorrectly deserialize
+# this null and get marshaling errors or even crashes - see issue #10417.
+# This test reproduces #10417, but not always - run with "--count" to
+# reproduce failures.
+@pytest.mark.xfail(reason="Issue #10417")
+def test_filtering_null_map_with_subscript(cql, test_keyspace):
+    schema = 'p text primary key, m map<int, int>'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (p) VALUES ('dog')")
+        assert list(cql.execute(f"SELECT p FROM {table} WHERE m[2] = 3 ALLOW FILTERING")) == []

--- a/test/cql-pytest/test_filtering.py
+++ b/test/cql-pytest/test_filtering.py
@@ -11,6 +11,7 @@
 # is or isn't necessary.
 
 import pytest
+import re
 from util import new_test_table
 from cassandra.protocol import InvalidRequest
 from cassandra.connection import DRIVER_NAME, DRIVER_VERSION
@@ -170,12 +171,21 @@ def test_filtering_with_in_relation(cql, test_keyspace, cassandra_bug):
         assert set(res) == set([(2,3,4,5), (4,5,6,7)])
 
 # Test that subscripts in expressions work as expected. They should only work
-# on map columns, and must have the correct type.
+# on map columns, and must have the correct type. Test that they also work
+# as expected for null or unset subscripts.
+# Cassandra considers the null subscript 'm[null]' to be an invalid request.
+# In Scylla we decided to it differently (we think better): m[null] is simply
+# a null, so the filter 'WHERE m[null] = 2' is not an error - it just doesn't
+# match anything. This is more consistent with our usual null handling
+# (null[2] and null < 2 are both defined as returning null), and will also
+# allow us in the future to support non-constant subscript - for example m[a]
+# where the column a can be null for some rows and non-null for other rows.
+# Because we decided that our behavior is better than Cassandra's, this test
+# fails on Cassandra and is marked with cassandra_bug.
 # This test is a superset of test test_null.py::test_map_subscript_null which
 # tests only the special case of a null subscript.
 # Reproduces #10361
-@pytest.mark.xfail(reason="Issue #10361")
-def test_filtering_with_subscript(cql, test_keyspace):
+def test_filtering_with_subscript(cql, test_keyspace, cassandra_bug):
     with new_test_table(cql, test_keyspace,
             "p int, m1 map<int, int>, m2 map<text, text>, s set<int>, PRIMARY KEY (p)") as table:
         # Check for *errors* in subscript expressions - such as wrong type or
@@ -187,24 +197,28 @@ def test_filtering_with_subscript(cql, test_keyspace):
         with pytest.raises(InvalidRequest, match='cannot be used as a map'):
             cql.execute(f"SELECT p FROM {table} WHERE s[2] = 3 ALLOW FILTERING")
         # A wrong type is passed for the subscript is not allowed
-        with pytest.raises(InvalidRequest, match='key\(m1\)'):
+        with pytest.raises(InvalidRequest, match=re.escape('key(m1)')):
             cql.execute(f"select p from {table} where m1['black'] = 2 ALLOW FILTERING")
-        with pytest.raises(InvalidRequest, match='key\(m2\)'):
+        with pytest.raises(InvalidRequest, match=re.escape('key(m2)')):
             cql.execute(f"select p from {table} where m2[1] = 2 ALLOW FILTERING")
-        # A "null" is not allowed as a key (reproduces #10361)
-        with pytest.raises(InvalidRequest, match='Unsupported null map key for column m1'):
-            cql.execute(f"select p from {table} where m1[null] = 2 ALLOW FILTERING")
-        with pytest.raises(InvalidRequest, match='Unsupported null map key for column m2'):
-            cql.execute(f"select p from {table} where m2[null] = 'hi' ALLOW FILTERING")
+        # See discussion of m1[null] above. Reproduces #10361, and fails
+        # on Cassandra (Cassandra deliberately returns an error here -
+        # an InvalidRequest with "Unsupported null map key for column m1"
+        assert list(cql.execute(f"select p from {table} where m1[null] = 2 ALLOW FILTERING")) == []
+        assert list(cql.execute(f"select p from {table} where m2[null] = 'hi' ALLOW FILTERING")) == []
         # Similar to above checks, but using a prepared statement. We can't
         # cause the driver to send the wrong type to a bound variable, so we
         # can't check that case unfortunately, but we have a new UNSET_VALUE
         # case.
         stmt = cql.prepare(f"select p from {table} where m1[?] = 2 ALLOW FILTERING")
-        with pytest.raises(InvalidRequest, match='Unsupported null map key for column m1'):
-            cql.execute(stmt, [None])
-        with pytest.raises(InvalidRequest, match='Unsupported unset map key for column m1'):
-            cql.execute(stmt, [UNSET_VALUE])
+        assert list(cql.execute(stmt, [None])) == []
+        # The expression m1[UNSET_VALUE] should be an error, but because the
+        # table is empty, we do not actually need to evaluate the expression
+        # and the error might might not be caught. So this test is commented
+        # out. We'll do it below, after we add some data to ensure that the
+        # expression does need to be evaluated.
+        #with pytest.raises(InvalidRequest, match='Unsupported unset map key for column m1'):
+        #    cql.execute(stmt, [UNSET_VALUE])
 
         # Finally, check for sucessful filtering with subscripts. For that we
         # need to add some data:
@@ -221,8 +235,9 @@ def test_filtering_with_subscript(cql, test_keyspace):
         # earlier when there was no data in the table. Now there is, and
         # the scan brings up several rows, it may exercise different code
         # paths.
-        with pytest.raises(InvalidRequest, match='Unsupported null map key for column m1'):
-            cql.execute(f"select p from {table} where m1[null] = 2 ALLOW FILTERING")
+        assert list(cql.execute(f"select p from {table} where m1[null] = 2 ALLOW FILTERING")) == []
+        with pytest.raises(InvalidRequest, match='Unsupported unset map key for column m1'):
+            cql.execute(stmt, [UNSET_VALUE])
 
 # Beyond the tests of map subscript expressions above, also test what happens
 # when the expression is fine (e.g., m[2] = 3) but the *data* itself is null.

--- a/test/cql-pytest/test_filtering.py
+++ b/test/cql-pytest/test_filtering.py
@@ -230,7 +230,6 @@ def test_filtering_with_subscript(cql, test_keyspace):
 # this null and get marshaling errors or even crashes - see issue #10417.
 # This test reproduces #10417, but not always - run with "--count" to
 # reproduce failures.
-@pytest.mark.xfail(reason="Issue #10417")
 def test_filtering_null_map_with_subscript(cql, test_keyspace):
     schema = 'p text primary key, m map<int, int>'
     with new_test_table(cql, test_keyspace, schema) as table:

--- a/test/cql-pytest/test_null.py
+++ b/test/cql-pytest/test_null.py
@@ -159,9 +159,16 @@ def test_filtering_null_comparison_no_filtering(cql, table1):
     with pytest.raises(InvalidRequest, match='ALLOW FILTERING'):
         cql.execute(f"SELECT c FROM {table1} WHERE p='x' AND m CONTAINS KEY NULL")
 
-# Test that null subscript m[null] is caught as the appropriate invalid-
-# request error, and not some internal server error or crash as we had in
-# #10361, #10399 and #10417.
+# Cassandra considers the null subscript 'm[null]' to be an invalid request.
+# In Scylla we decided to it differently (we think better): m[null] is simply
+# a null, so the filter 'WHERE m[null] = 3' is not an error - it just doesn't
+# match anything. This is more consistent with our usual null handling (null[2]
+# and null < 2 are both defined as returning null), and will also allow us
+# in the future to support non-constant subscript - for example m[a] where
+# the column a can be null for some rows and non-null for other rows.
+# Before we implemented the above decision, we had multiple bugs in this case,
+# resulting in bizarre errors and even crashes (see #10361, #10399 and #10417).
+#
 # Because this test uses a shared table (table1), then depending on how it's
 # run, it sometimes sees an empty table and sometimes a table with data
 # (and null values for the map m...), so this test mixes several different
@@ -169,9 +176,6 @@ def test_filtering_null_comparison_no_filtering(cql, table1):
 # by test_filtering.py::test_filtering_with_subscript and
 # test_filtering.py::test_filtering_null_map_with_subscript so this test
 # should eventually be deleted.
-@pytest.mark.xfail(reason="Issue #10361, #10399")
-def test_map_subscript_null(cql, table1):
-    with pytest.raises(InvalidRequest, match='null'):
-        cql.execute(f"SELECT p FROM {table1} WHERE m[null] = 3 ALLOW FILTERING")
-    with pytest.raises(InvalidRequest, match='null'):
-        cql.execute(cql.prepare(f"SELECT p FROM {table1} WHERE m[?] = 3 ALLOW FILTERING"), [None])
+def test_map_subscript_null(cql, table1, cassandra_bug):
+    assert list(cql.execute(f"SELECT p FROM {table1} WHERE m[null] = 3 ALLOW FILTERING")) == []
+    assert list(cql.execute(cql.prepare(f"SELECT p FROM {table1} WHERE m[?] = 3 ALLOW FILTERING"), [None])) == []

--- a/test/cql-pytest/test_null.py
+++ b/test/cql-pytest/test_null.py
@@ -169,7 +169,7 @@ def test_filtering_null_comparison_no_filtering(cql, table1):
 # by test_filtering.py::test_filtering_with_subscript and
 # test_filtering.py::test_filtering_null_map_with_subscript so this test
 # should eventually be deleted.
-@pytest.mark.xfail(reason="Issue #10361, #10399, 10417")
+@pytest.mark.xfail(reason="Issue #10361, #10399")
 def test_map_subscript_null(cql, table1):
     with pytest.raises(InvalidRequest, match='null'):
         cql.execute(f"SELECT p FROM {table1} WHERE m[null] = 3 ALLOW FILTERING")

--- a/test/cql-pytest/test_null.py
+++ b/test/cql-pytest/test_null.py
@@ -160,8 +160,16 @@ def test_filtering_null_comparison_no_filtering(cql, table1):
         cql.execute(f"SELECT c FROM {table1} WHERE p='x' AND m CONTAINS KEY NULL")
 
 # Test that null subscript m[null] is caught as the appropriate invalid-
-# request error, and not some internal server error as we had in #10361.
-@pytest.mark.xfail(reason="Issue #10361")
+# request error, and not some internal server error or crash as we had in
+# #10361, #10399 and #10417.
+# Because this test uses a shared table (table1), then depending on how it's
+# run, it sometimes sees an empty table and sometimes a table with data
+# (and null values for the map m...), so this test mixes several different
+# concerns and problems. The same problems are better covered separately
+# by test_filtering.py::test_filtering_with_subscript and
+# test_filtering.py::test_filtering_null_map_with_subscript so this test
+# should eventually be deleted.
+@pytest.mark.xfail(reason="Issue #10361, #10399, 10417")
 def test_map_subscript_null(cql, table1):
     with pytest.raises(InvalidRequest, match='null'):
         cql.execute(f"SELECT p FROM {table1} WHERE m[null] = 3 ALLOW FILTERING")

